[RNTuple] Implemented dynamic class definitions, type annotations, class hierarchies, renamed classes and files

sathabbott opened the pull request.

sathabbott committed:
Merging WIP RNTuple initial dev into upstream dynamic updates (#2)

* [WIP] RNTuple initial dev

update unpack to use unsigned integers (could be pointless)

add debug statements, len(buffer) -> buffer.__len__()

add rntuple reader prototype + test root file

improved debug messages

TKeyList temp bug fix (will need to return to it) + some debug updates

checkpoint, can read file that contains RNTuple but not read RNTuple stuff yet

Add RNTuple & RNTupleAnchor classes

[WIP] RNTupleEnvelope initialized

remove obsolete file

pre-nox checkpoint. working, can print pages. use next commit w/ nox changes though.

[WIP] handle initial nox errors

ignore print statements in reader test file

[CHECKPOINT] passed nox: removed debugging comments & print statements

[CHECKPOINT] minor bug fixes

* Update pyproject.toml [resolving merge]

* resolving conflicts for PR#2

sathabbott committed:
fix nox file for rntuple reader ignore print

sathabbott committed:
move test files

sathabbott committed:
minor bug fix

sathabbott committed:
renamed objects (RNTuple -> R)

sathabbott committed:
[RNTuple] implement dynamic class definitions, type annotations, class hierarchies

sathabbott committed:
fix reader to not use debugging code

sathabbott commented:
TODO still:
- Deserialize Header Envelope
- Use Header Envelope to deserialize page data into cpp types

sathabbott commented:
rootfilespec class diagrams

![packages_rootfilespec](https://github.com/user-attachments/assets/6248c27c-d1f2-437b-9da3-39dcdb9af7e9)
[packages_rootfilespec.pdf](https://github.com/user-attachments/files/19797302/packages_rootfilespec.pdf)
[classes_rootfilespec.pdf](https://github.com/user-attachments/files/19797306/classes_rootfilespec.pdf)
![classes_rootfilespec](https://github.com/user-attachments/assets/404afba3-6897-4b32-9c59-6656f9a9078f)

sathabbott committed:
fix lint errors

sathabbott committed:
separate REnvelopeLink & RLocator, fix typing

sathabbott commented:
rootfilespec class diagrams 

[packages_rootfilespec.pdf](https://github.com/user-attachments/files/19842051/packages_rootfilespec.pdf)
![packages_rootfilespec](https://github.com/user-attachments/assets/5371adf2-184f-4f33-a54e-023c1154a537)

[classes_rootfilespec.pdf](https://github.com/user-attachments/files/19842056/classes_rootfilespec.pdf)
![classes_rootfilespec](https://github.com/user-attachments/assets/bd22f04a-819a-4b47-9f54-dc8240b6894a)

nsmith- commented on src/rootfilespec/bootstrap/TFile.py:
```diff
@@ -123,15 +123,42 @@ class ROOTFile_header_v622_large(ROOTSerializable):
+    """A class representing a ROOT file.
+    Binary Spec: https://root.cern.ch/doc/master/classTFile.html
+
+    Attributes:
```
Granted, I have not yet implemented this on many of the existing classes. I'll do that in a separate PR soon

nsmith- reviewed (changes requested):
Overall very solid. There's a few organizational things to touch up and some dead code.
We should remove the `RNTuple.root` file from this repository and add it to scikithep test data (if not already there)
It's generally a good idea to avoid putting large binaries in a repo if there is an alternative.

sathabbott commented on rntuple_reader.py:
```diff
@@ -0,0 +1,146 @@
+if __name__ == "__main__":
+    initial_read_size = 512
+    # path = Path("../TTToSemiLeptonic_UL18JMENanoAOD-zstd.root")
+    path = Path("tests/RNTuple.root")
```
I got it from [rntviewer](https://codeberg.org/silverweed/rntviewer/src/branch/master/testdata) so i could compare against the graphical hex dump. 

it might be the same one as [this one](https://github.com/scikit-hep/scikit-hep-testdata/blob/main/src/skhep_testdata/data/ntpl001_staff_rntuple_v1-0-0-0.root)(?) I'll download it and check when i can later.

sathabbott replied:
It is not the same file, I am adding the file to the scikit-hep test data and deleting it from our repo.

sathabbott replied:
[scikit-hep test data PR to add rntviewer testdata](https://github.com/scikit-hep/scikit-hep-testdata/pull/174)

sathabbott committed:
raise error if unfinished get_header() function called

Co-authored-by: Nicholas Smith <nick.smith@cern.ch>

sathabbott committed:
rename and reorganize envelope files

sathabbott committed:
streamline FooterEnvelope.get_pagelilst, lint

sathabbott commented on src/rootfilespec/bootstrap/REnvelope.py:
```diff
@@ -0,0 +1,227 @@
+
+        return (headerChecksum, clusterSummaries, pageLocations), buffer
+
+    def get_pages(self, fetch_data: DataFetcher):
```
i agree, it will eventually be replaced by the `RFrame.PageLocations_Clusters.find_page` method, but that method is untested currently. Will update later.

sathabbott committed:
auto generate read method for REnvelopeLink

sathabbott committed:
streamline REnvelopeLink

sathabbott committed:
finish streamlining REnvelopeLink

sathabbott commented on src/rootfilespec/bootstrap/RFrame.py:
```diff
@@ -0,0 +1,390 @@
+        #### Read the page from the buffer
+        page, buffer = RPage.read(buffer)
+
+        # TODO: compression
```
[According to the RNTuple documentation](https://github.com/root-project/root/blob/v6-34-00-patches/tree/ntuple/v7/doc/BinaryFormatSpecification.md#compression-block):
`RPage`s and `REnvelope`s are wrapped in compression blocks. In order to deserialize a page or an envelope, its compressed and uncompressed size needs to be known.

I haven't decided on how to implement it exactly for `RPage`, but I was leaning towards `RPageDescription`, which should match how `REnvelopeLink` steers the compression (for now).

sathabbott committed:
streamlining RFrame

sathabbott commented on src/rootfilespec/bootstrap/TFile.py:
```diff
@@ -91,16 +91,16 @@ class ROOTFile_header_v622_small(ROOTSerializable):
-    fNbytesFree: Annotated[int, Fmt(">i")]
-    nfree: Annotated[int, Fmt(">i")]
-    fNbytesName: Annotated[int, Fmt(">i")]
+    fBEGIN: Annotated[int, Fmt(">I")]
```
It isn't stated directly, but based on the documentation, they seem unsigned:

https://root.cern.ch/doc/master/header.html

sathabbott replied:
looking at the [source code](https://root.cern.ch/doc/master/TFile_8h_source.html#l00162), they use signed integers, so we should too.

sathabbott commented on src/rootfilespec/bootstrap/TFile.py:
```diff
@@ -26,7 +26,7 @@ class VersionInfo(ROOTSerializable):
     @classmethod
     def read(cls, buffer: ReadBuffer):
-        (version,), buffer = buffer.unpack(">i")
+        (version,), buffer = buffer.unpack(">I")
```
I didn't see one personally, but based on the documentation, they seem unsigned:

https://root.cern.ch/doc/master/header.html

sathabbott committed:
use docstring for TKeyList

sathabbott committed:
use signed integers to match root source code

sathabbott committed:
update TSring docstring

sathabbott committed:
remove unused code

sathabbott committed:
remove type annotation from get_KeyList method

sathabbott committed:
remove testdata

sathabbott committed:
move page locations/descriptions into separate file

sathabbott committed:
update docstrings, move docstrings to fields

sathabbott commented:
All done, all comments implemented!

sathabbott commented:
On the last commit I went through and moved a bunch of docstrings inline, and cleaned up the class docstrings. I think you should look over the vanilla ROOT changes before approving this commit to make sure there isn't anything you want to fix.

There are a couple files I didn't update the docstrings for (left you some fun!)

sathabbott committed:
lint

nsmith- commented on src/rootfilespec/bootstrap/envelopebase.py:
```diff
@@ -0,0 +1,138 @@
+        #   Envelope Bytes Read  = buffer.relpos - payload_start_pos
+
+        #### Get the checksum (appended to envelope when writing to disk)
+        (checksum,), buffer = buffer.unpack("<Q")  # Last 8 bytes of the envelope
```
Eventually we will want to check the checksum here, and then there will not be any purpose in storing it in the dataclass (since it is dependent on the content)

nsmith- committed:
Few adjustments

nsmith- reviewed (approved):
Looks great now!
I'll do some re-arrangement of the files into an `rntuple` submodule in a follow-up PR
